lib/spack/spack/solver/asp.py

```diff
@@ -2016,33 +2022,74 @@ def _develop_specs_from_env(spec, env):
     spec.constrain(dev_info['spec'])
 
 
-#
-# These are handwritten parts for the Spack ASP model.
-#
-def solve(specs, dump=(), models=0, timers=False, stats=False, tests=False,
-          reuse=False):
-    """Solve for a stable model of specs.
-
-    Arguments:
-        specs (list): list of Specs to solve.
-        dump (tuple): what to dump
-        models (int): number of models to search (default: 0)
+class Solver(object):
```

Comment:
Thoughts on this class after our webex meeting:

The `Solver` class constructor should contain any calls to `spack.config` to get concretizer defaults -- any changes to the config should not propagate to an existing object. Client code (`Spec.concretize`, `Environment.concretize`) should instantiate a new `Solver` object if it needs a new one, or it can explicitly set attributes to control the behavior of an existing `Solver` object.

The `Solver` object should also come with a driver. For now there's only one driver, but we should instantiate a `PyclingoDriver` at `Solver` instantiation time.

The `out` argument for `PyclingoDriver` should move to the `solve` method. All output-related arguments (`show`, `models`, `timers`, `stats`) should be arguments to the solve method, because they're relevant to a specific solve, not to the solver object.

The `SolverSetup` object has to be specific to a given solve, because it maintains state. So `Solver.solve` instantiates a `SolverSetup` and calls `driver.solve` with that setup object.

I also think the driver should become a required constructor parameter for the `SolverSetup` object -- we're never going to use the same SolverSetup object with different drivers.

Does that all make sense and allow for the simplification of the interface that you were looking for with the changes here? Trying to go with a rules-based description rather than code suggestions so the logic behind it is clearer, and whatever we agree on we can put into the docstrings.

Reply:
I did all of these *except* for this one:

> I also think the driver should become a required constructor parameter for the `SolverSetup` object -- we're never going to use the same `SolverSetup` object with different drivers.

Mainly because the `SolverSetup` can be used *without* a driver -- as it is in `spack diff` to create facts for diffing.  I don't think we should construct a driver there and couple the two classes that tightly, when all we want is to get facts for a set of specs.